File: jeffrey-microscope/profiles/heap-dump/src/main/java/cafe/jeffrey/profile/heapdump/model/HeapDumpInitProgress.java
/*
 * Jeffrey
 * Copyright (C) 2026 Petr Bouda
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package cafe.jeffrey.profile.heapdump.model;

import java.util.List;

/**
 * Live progress of a server-side heap-dump initialization run, polled by the
 * frontend while the pipeline executes in the background.
 *
 * @param state        {@code idle} (no run for this profile), {@code running},
 *                     {@code completed} or {@code failed}
 * @param errorCode    machine-readable error code when {@code failed}, else {@code null}
 * @param errorMessage human-readable failure description when {@code failed}, else {@code null}
 * @param stages       per-stage live statuses in pipeline order; empty when {@code idle}
 */
public record HeapDumpInitProgress(
        String state,
        String errorCode,
        String errorMessage,
        List<HeapDumpInitStageProgress> stages
) {

    public static final String STATE_IDLE = "idle";

    public static final String STATE_RUNNING = "running";

    public static final String STATE_COMPLETED = "completed";

    public static final String STATE_FAILED = "failed";

    public static HeapDumpInitProgress idle() {
        return new HeapDumpInitProgress(STATE_IDLE, null, null, List.of());
    }
}

File: jeffrey-microscope/pages-microscope/src/services/api/model/HeapDumpInitProgress.ts
/*
 * Jeffrey
 * Copyright (C) 2026 Petr Bouda
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { SubPhaseTiming } from '@/services/api/model/InitPipelineResult';

export interface HeapDumpInitStageProgress {
  id: string;
  status: 'pending' | 'in_progress' | 'completed' | 'failed' | 'skipped';
  durationMs: number | null;
  /** Backend-measured milliseconds spent so far while in_progress, else null. */
  elapsedMs: number | null;
  subPhases: SubPhaseTiming[] | null;
}

export default interface HeapDumpInitProgress {
  state: 'idle' | 'running' | 'completed' | 'failed';
  errorCode: string | null;
  errorMessage: string | null;
  stages: HeapDumpInitStageProgress[];
}
